commands/importer/validate.go

```diff
@@ -0,0 +1,904 @@
+
+	return nil
+}
+
+func (v *Validator) validateLine(info ImportFileInfo, line LineImportData) error {
+	var err error
+
+	// make sure the file starts with a version
+	if info.CurrentLine == 1 && line.Type != "version" {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("first line has the wrong type: expected \"version\", got %q", line.Type),
+		}); err != nil {
+			return err
+		}
+	}
+
+	switch line.Type {
+	case "version":
+		err = v.validateVersion(info, line)
+	case "scheme":
+		err = v.validateScheme(info, line)
+	case "team":
+		err = v.validateTeam(info, line)
+	case "channel":
+		err = v.validateChannel(info, line)
+	case "user":
+		err = v.validateUser(info, line)
+	case "post":
+		err = v.validatePost(info, line)
+	case "direct_channel":
+		err = v.validateDirectChannel(info, line)
+	case "direct_post":
+		err = v.validateDirectPost(info, line)
+	case "emoji":
+		err = v.validateEmoji(info, line)
+	default:
+		err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			FieldName:      "type",
+			Err:            fmt.Errorf("unknown import type %q", line.Type),
+		})
+	}
+
+	return err
+}
+
+func (v *Validator) validateVersion(info ImportFileInfo, line LineImportData) (err error) {
+	if info.CurrentLine != 1 {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("version info must be the first line of the file"),
+		}); err != nil {
+			return err
+		}
+	}
+
+	if line.Version == nil {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("version must not be null or missing"),
+		}); err != nil {
+			return err
+		}
+	} else if *line.Version != 1 {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("version must not be 1"),
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateScheme(info ImportFileInfo, line LineImportData) (err error) {
+	ive := validateNotNil(info, "scheme", line.Scheme, func(data SchemeImportData) *ImportValidationError {
+		appErr := validateSchemeImportData(&data)
+		if appErr != nil {
+			return &ImportValidationError{
+				ImportFileInfo: info,
+				FieldName:      "scheme",
+				Err:            appErr,
+			}
+		}
+
+		if data.Name != nil {
+			if existing, ok := v.schemes[*data.Name]; ok {
+				return &ImportValidationError{
+					ImportFileInfo: info,
+					FieldName:      "scheme",
+					Err:            fmt.Errorf("duplicate entry, previous was in line: %d", existing.CurrentLine),
+				}
+			}
+			v.schemes[*data.Name] = info
+		}
+
+		return nil
+	})
+	if ive != nil {
+		return v.onError(ive)
+	}
+
+	return nil
+}
+
+func (v *Validator) validateTeam(info ImportFileInfo, line LineImportData) (err error) {
+	ive := validateNotNil(info, "team", line.Team, func(data TeamImportData) *ImportValidationError {
+		appErr := validateTeamImportData(&data)
```

Comment:
Any reason to be passing pointers to these validate functions? We just validated that it's non nil, so it should be fine to just pass values.

Reply:
The reason is that the validate functions take pointers. I kept it to maintain compatibility to the functions in the server.

Reply:
But they seem to be copied over right? Are these somehow shared? No issues either ways. Fine to keep it like this.

Reply:
Let's keep it for now. I'd like to update the validators themselves in a later PR, so this can be reflected.